### DISCUSSION TITLE
v2ray的h2c能不能支持Apache2

### DISCUSSION THREAD

**Discussion by ghost** [v2fly/v2ray-core#1685] [Q&A]
## 你正在使用哪个版本的 V2Ray？

4.44

## 你的使用场景是什么？

通过v2ray访问网站

## 你看到的异常现象是什么？

v2ray没有反应，网页无法打开

## 你期待看到的正常表现是怎样的？

v2ray正常工作，正常代理网页

## 请附上你的配置

**服务端配置：**

```javascript
// 在这里附上服务器端配置文件
Apache2: 
ProxyPass h2c://(v2ray的地址）
ProxyPassReverse h2c://(v2ray的地址）

v2ray: 
和模板里的caddy h2c配置差不多
```

**客户端配置：**

```javascript
// 在这里附上客户端配置
和caddy h2c一样

```

## 请附上出错时软件输出的错误日志

提示timeout

**服务器端错误日志：**

```javascript
// 在这里附上服务器端日志

```

**客户端错误日志：**

```javascript
// 在这里附上客户端日志

```

## 请附上访问日志

日志是空的，没有访问记录

```javascript
// 在这里附上服务器端日志

```

## 其它相关的配置文件（如 Nginx）和相关日志
Apache2日志是空的，没有访问的记录

## 如果 V2Ray 服务运行异常，请附上 journal 日志
v2ray日志没有显示我的设备访问了它

**Comment by AkinoKaede** [comment]
I am not good at using Apache 2, but I found some documentation.
<https://httpd.apache.org/docs/trunk/mod/mod_proxy_http2.html>

**Reply by ghost** [reply]
Yeah, I configured Apache2 like this, with proxy_http2 module enabled: 
<Location "(v2ray http2 path)">
ProxyPass h2c://(v2ray local address with port and path)
ProxyPassReverse h2c://(v2ray local address with port and path)
<\/Location>

then I set v2ray's network to h2 and security to none, both v2ray and apache2 is restarted after these configuration changes, but it doesn't seem to be working. I was wondering if there are some options missing in apache2 configuration, such as nocanon or some other headers.

**Reply by lxhao61** [reply]
> Yeah, I configured Apache2 like this, with proxy_http2 module enabled: <Location "(v2ray http2 path)"> ProxyPass h2c://(v2ray local address with port and path) ProxyPassReverse h2c://(v2ray local address with port and path) </Location>
> 
> then I set v2ray's network to h2 and security to none, both v2ray and apache2 is restarted after these configuration changes, but it doesn't seem to be working. I was wondering if there are some options missing in apache2 configuration, such as nocanon or some other headers.

看了Apache模块资料，觉得需要同时启用mod_http2与mod_proxy_http2模块。
1、伪装代理网站配置参数[Protocols h2 http/1.1](https://httpd.apache.org/docs/trunk/mod/mod_http2.html)。
2、反向代理h2c配置中删除“ProxyPassReverse h2c://(v2ray local address with port and path)”此项。

**Reply by ghost** [reply]
我试了一下你说的配置，网页可以上了，但在某些情况下还是不行，比如发推上传图片是上传不了的。

**Reply by lxhao61** [reply]
大方向应该对了，最终解决我已帮不了你了：因为我不使用这个，故对Apache2无深入研究。就目前情况来看，应该加入些优化参数来解决；可参考Apache2反代WebSocket或Nginx反代gRPC配置示例及原理。
